### PR TITLE
prov/tcp Adding auto progress support

### PR DESCRIPTION
**Pull request by vkrishna**
Signed-off-by: Venkata Krishna Nimmagadda <venkata.krishna.nimmagadda@intel.com>

### CODE REVIEW

File: prov/tcp/src/tcpx_ep.c [hunk 1]
```diff
@@ -75,7 +77,9 @@ static ssize_t tcpx_recvmsg(struct fid_ep *ep, const struct fi_msg *msg,
 	recv_entry->context = msg->context;
 	recv_entry->done_len = 0;
 
+	fastlock_acquire(&tcpx_ep->queue_lock);
 	dlist_insert_tail(&recv_entry->entry, &tcpx_ep->rx_queue);
+	fastlock_release(&tcpx_ep->queue_lock);
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: This looks like a real problem separate from progress.  A fix for synchronization issues should go into its own patch.

File: prov/tcp/src/tcpx_ep.c [hunk 2]
```diff
@@ -75,7 +77,9 @@ static ssize_t tcpx_recvmsg(struct fid_ep *ep, const struct fi_msg *msg,
 	recv_entry->context = msg->context;
 	recv_entry->done_len = 0;
 
+	fastlock_acquire(&tcpx_ep->queue_lock);
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: It would be better to just acquire/release the lock once.

File: prov/tcp/src/tcpx_ep.c [hunk 3]
```diff
@@ -136,8 +141,10 @@ static ssize_t tcpx_sendmsg(struct fid_ep *ep, const struct fi_msg *msg,
 	data_len = ofi_total_iov_len(msg->msg_iov, msg->iov_count);
 
 	if (flags & FI_INJECT) {
-		if (data_len > TCPX_MAX_INJECT_SZ)
-			return -FI_EINVAL;
+		if (data_len > TCPX_MAX_INJECT_SZ) {
+			ret = -FI_EINVAL;
+			goto err;
+		}
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: This should be an assert() rather than a runtime check.  The app has the inject size at initialization and should ensure that it limits transfers appropriately.

File: prov/tcp/src/tcpx_progress.c [hunk 4]
```diff
@@ -42,8 +42,32 @@
 #include <ofi_util.h>
 #include <ofi_iov.h>
 
+void auto_prog_mgr_close(struct auto_prog_mgr *mgr)
```
Comment (shefty) [inline prov/tcp/src/tcpx_progress.c]: Please use prefix tcpx_ in all functions for this provider to avoid name space collisions and allow easy identification of code ownership.

File: prov/tcp/src/tcpx_progress.c [hunk 5]
```diff
@@ -173,40 +197,260 @@ static void process_rx_pe_entry(struct tcpx_pe_entry *pe_entry)
...
+	struct fid_list_entry *item;
+
+	dlist_foreach(&mgr->ep_list,entry) {
+		item = container_of(entry, struct fid_list_entry, entry);
+		ep = container_of(item->fid, struct tcpx_ep,
+				  util_ep.ep_fid.fid);
+
+		fastlock_acquire(&ep->queue_lock);
+		progress_tx_queue(ep);
+		fastlock_release(&ep->queue_lock);
```
Comment (shefty) [inline prov/tcp/src/tcpx_progress.c]: Move locks inside progress calls

File: prov/tcp/src/tcpx_ep.c (hunk 1 above)
Comment (vkrishna) [reply prov/tcp/src/tcpx_ep.c]: Created a separate commit. Thanks

**Comment by shefty** [comment]
How is this PR related to #3975? Does that PR layer over this one?

**Comment by vkrishna** [comment]
Actually this was created earlier just for auto progress. Then I reworked manual progress to maximize code reuse with this PR.  Once that is reviewed, I wanted to re-base this on top of that PR. Please either  label this WIP or close this PR until manual progress is complete.

W.r.t the comments on PR #3975, I do agree that this approach is not optimal.  For manual progress using CQ util code may be better.
